include/pmacc/lockstep/WorkerCfg.hpp

```diff
@@ -0,0 +1,117 @@
+/* Copyright 2022 Rene Widera
+ *
+ * This file is part of PMacc.
+ *
+ * PMacc is free software: you can redistribute it and/or modify
+ * it under the terms of either the GNU General Public License or
+ * the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * PMacc is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with PMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include "pmacc/lockstep/Worker.hpp"
+#include "pmacc/traits/GetNumWorkers.hpp"
+#include "pmacc/types.hpp"
+
+#include "pmacc/math/vector/compile-time/Vector.hpp"
+
+
+namespace pmacc::lockstep
+{
+    namespace detail
+    {
+        struct WorkerCfgAssume1DThreads
+        {
+            template<typename T_WorkerCfg, typename T_Acc>
+            HDINLINE static auto getWorker(T_Acc const& acc)
+            {
+                return T_WorkerCfg::getWorkerAssume1DThreads(acc);
+            }
+        };
+    } // namespace detail
+    /** Configuration of worker used for a lockstep kernel
+     *
+     * @tparam T_numWorkers Number of workers which are used to execute this functor.
+     *
+     * @attention: The real number of workers used for the lockstep kernel depends on the alpaka backend and will
+     * be adjusted by this class via the trait pmacc::traits::GetNumWorkers.
+     */
+    template<uint32_t T_numWorkers>
+    struct WorkerCfg
+    {
+        friend struct detail::WorkerCfgAssume1DThreads;
+
+        //! number of workers
+        static constexpr uint32_t numWorkers = pmacc::traits::GetNumWorkers<T_numWorkers>::value;
```

Comment:
No, depending on the used cupla/accelerator the trait is giving other values back. For OMP2block the trait will return`1`.
The reason is that some alpaka accelerator has a restriction for the block size.
Before this PR the user always called this trait. Now the call is moved into `WorkerCfg` because calling this trait is an essential part of lockstep programming.

Reply:
Are you sure about this? The trait I linked to has a default value for the template parameter `T_ACC` and no second template parameter is passed here. To get another specialization, you would need to pass an accelerator type, e.g. `GetNumWorkers<T_numWorkers, AccOmp2Blocks<Idx, Dim>>`.

Reply:
Yes I am sure.

https://github.com/ComputationalRadiationPhysics/picongpu/blob/59e9b53605f9a5c1bf271eeb055bc74370a99052/include/pmacc/traits/GetNumWorkers.hpp#L41-L52

The default will be used and then the compiler selects the specialized implementation. If this did not work you would not be able to run `serial` or `omp2b` as PIConGPU backend because both backends have the limitation that only one thread per block is allowed.

Reply:
Ahh, so this is because `cupla::AccThreadSeq` is an alias for the default accelerator configured via cmake. I am not yet familiar enough with cupla: to me `AccThreadSeq` read like `::alpaka::AccCpuSerial`, I thought it was a fixed alias to a single alpaka backend.